Ability to backup snapshots for managed storage

syed opened the pull request:
* Provides functionality to do archvie of snapshots
* Adds the LVM-less SR

mike-tutkowski commented on api/src/org/apache/cloudstack/api/command/user/snapshot/CreateSnapshotCmd.java:
```diff
@@ -224,8 +224,8 @@ public void execute() {
 
-        if (locationType != null && locationType >= 0 && locationType < Snapshot.LocationType.values().length) {
-            locationTypeToReturn = Snapshot.LocationType.values()[locationType];
+        if (locationType != null && locationType >= 0 && locationType <= Snapshot.LocationType.values().length) {
```
If the user can only specify 1 or 2 (which is currently the case), then locationType >= 0 should be locationType >= 1.

Also, the description for locationType should specify the text 'Standard' instead of 'Primary'.

Currently says: "Currently applicable only for managed storage. " +
            "Valid location types: 1 = 'primary'; 2 = 'archive'. Default = 1."

Should say: "Currently applicable only for managed storage. " +
            "Valid location types: 1 = 'standard'; 2 = 'archive'. Default = 1."

syed replied:
I thought I fixed that. I did exactly what you decribed here Mike. It looke like somehow it got missed. I will fix that.

mike-tutkowski commented on engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:
```diff
@@ -207,13 +205,9 @@ public void copyAsync(DataObject srcData, DataObject destData, Host destHost, As
-                    LOGGER.warn(errMsg);
-
-                    throw new UnsupportedOperationException(errMsg);
+                if (canHandleDest) {
```
It looks like we have two "if (canHandleDest)" statements right in a row doing different things.

syed replied:
Same here. I was trying to rebase to the master which got ugly really quickly then I reverted back to the old code and somewhere in between I lot the commit. I will fix this

mike-tutkowski commented on api/src/org/apache/cloudstack/api/command/user/snapshot/CreateSnapshotCmd.java:
```diff
@@ -224,7 +224,7 @@ public void execute() {
         Snapshot.LocationType locationTypeToReturn = Snapshot.LocationType.values()[0];
 
-        if (locationType != null && locationType >= 0 && locationType <= Snapshot.LocationType.values().length) {
+        if (locationType != null && locationType >= 1 && locationType <= Snapshot.LocationType.values().length) {
```
Looks good.

Maybe you can update the description of locationType to change the word 'primary' to 'standard', too. Thanks

syed commented:
@mike-tutkowski I've removed the vhdoiscsi-sr and fixed a bunch of things. Can you review this now? I'm currently writing unit tests for these currently.

mike-tutkowski commented on core/src/com/cloud/storage/template/VhdProcessor.java:
```diff
@@ -91,27 +97,54 @@ public long getVirtualSize(File file) throws IOException {
     }
 
-    protected long getTemplateVirtualSize(File file) throws IOException {
+        protected long getTemplateVirtualSize(File file) throws IOException {
```
Looks like this is accidentally indented a bit.

mike-tutkowski commented on core/src/com/cloud/storage/template/VhdProcessor.java:
```diff
@@ -45,10 +49,12 @@
     private static final Logger s_logger = Logger.getLogger(VhdProcessor.class);
     StorageLayer _storage;
     private int vhdFooterSize = 512;
+    private int vhdCookieOffset = 8;
```
Hey Syed - Can you give me a bit of background on this file needing to be changed? Is this a bug fix that is being integrated into the work of this PR or is this file being changed because of the new functionality of this PR? Thanks!

syed replied:
@mike-tutkowski this was a fix that I had done for Swift to work with comrpessed templates. I've opened another PR which fixes this in Master. You can ignore these changes. I plan to remove them

mike-tutkowski commented on engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:
```diff
@@ -274,13 +281,74 @@ private void validate(SnapshotInfo snapshotInfo) {
+        return selectedScope;
+    }
+
+    private Scope getZoneScope(Scope destScope) {
```
Both srcScope and destScope are passed to this method (in pickCacheScopeForCopy), but the parameter here is called destScope, which might be a bit misleading in the case where srcScope is passed to this method.

mike-tutkowski commented on engine/storage/snapshot/src/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java:
```diff
@@ -93,7 +90,35 @@
+            return snapshotInfo;
+        }
+
+        // At this point the snapshot is either taken as a nati:0
```
Is this a typo? nati:0

mike-tutkowski commented:
Hey Syed - Aside from a few comments, this looks good to me.

I plan to pull this code down and run my Marvin tests against it to make sure there are no breakages there.

I should be able to report back on the results toward the end of the day.

syed commented:
Thanks @mike-tutkowski , I've addressed your comments and some minor stuff that I found. I have not tested this on a normal LVM-based SR. Would be interested in the results.

mike-tutkowski commented:
Sounds good

I fetched your latest code and am compiling it now.

syed commented:
@mike-tutkowski is it possible that you pull the code again? I found a small but when doing resign

mike-tutkowski commented:
Sure, no problem.

My first test suite had 1 failure out of 11 tests, but I don't think the failure has anything to do with your new code.

I can pull the new code and re-run that suite.

mike-tutkowski commented:
My first suite of tests, TestVolumes.py, has passed (11 tests). I ran it without making use of resigning logic.

I'm running this suite again making use of resigning logic.

mike-tutkowski commented:
Here's the stack trace I get:

SR_BACKEND_FAILURE_47The SR is not available [opterr=The SR is not available [opterr=no such volume group: VG_XenStorage-fff6f0d5-aba6-601f-2272-566c63427a1a]]
    at com.xensource.xenapi.Types.checkResponse(Types.java:2021)
    at com.xensource.xenapi.Connection.dispatch(Connection.java:395)
    at com.cloud.hypervisor.xenserver.resource.XenServerConnectionPool$XenServerConnection.dispatch(XenServerConnectionPool.java:457)
    at com.xensource.xenapi.PBD.plug(PBD.java:457)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.introduceAndPlugIscsiSr(CitrixResourceBase.java:2411)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.getIscsiSR(CitrixResourceBase.java:2362)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.getIscsiSR(CitrixResourceBase.java:2298)
    at com.cloud.hypervisor.xenserver.resource.XenServerStorageProcessor.resignature(XenServerStorageProcessor.java:177)
    at com.cloud.hypervisor.xenserver.resource.XenServerStorageProcessor.resignature(XenServerStorageProcessor.java:91)
    at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:69)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixStorageSubSystemCommandWrapper.execute(CitrixStorageSubSystemCommandWrapper.java:36)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixStorageSubSystemCommandWrapper.execute(CitrixStorageSubSystemCommandWrapper.java:30)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixRequestWrapper.execute(CitrixRequestWrapper.java:122)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.executeRequest(CitrixResourceBase.java:1683)
    at com.cloud.agent.manager.DirectAgentAttache$Task.runInContext(DirectAgentAttache.java:315)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
WARN  [c.c.h.x.r.XenServerStorageProcessor](DirectAgent-292:ctx-406b5eaf) (logid:32dbf175) Failed to resignature: com.cloud.utils.exception.CloudRuntimeException: Unable to create Iscsi SR  {targetIQN=iqn.2010-01.com.solidfire:u19v.root-21.46, target=10.10.8.112, SCSIid=36f47acc100000000753139760000002e} due to  SR_BACKEND_FAILURE_47The SR is not available [opterr=The SR is not available [opterr=no such volume group: VG_XenStorage-fff6f0d5-aba6-601f-2272-566c63427a1a]]
com.cloud.utils.exception.CloudRuntimeException: Unable to create Iscsi SR  {targetIQN=iqn.2010-01.com.solidfire:u19v.root-21.46, target=10.10.8.112, SCSIid=36f47acc100000000753139760000002e} due to  SR_BACKEND_FAILURE_47The SR is not available [opterr=The SR is not available [opterr=no such volume group: VG_XenStorage-fff6f0d5-aba6-601f-2272-566c63427a1a]]
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.getIscsiSR(CitrixResourceBase.java:2371)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.getIscsiSR(CitrixResourceBase.java:2298)
    at com.cloud.hypervisor.xenserver.resource.XenServerStorageProcessor.resignature(XenServerStorageProcessor.java:177)
    at com.cloud.hypervisor.xenserver.resource.XenServerStorageProcessor.resignature(XenServerStorageProcessor.java:91)
    at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:69)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixStorageSubSystemCommandWrapper.execute(CitrixStorageSubSystemCommandWrapper.java:36)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixStorageSubSystemCommandWrapper.execute(CitrixStorageSubSystemCommandWrapper.java:30)
    at com.cloud.hypervisor.xenserver.resource.wrapper.xenbase.CitrixRequestWrapper.execute(CitrixRequestWrapper.java:122)
    at com.cloud.hypervisor.xenserver.resource.CitrixResourceBase.executeRequest(CitrixResourceBase.java:1683)
    at com.cloud.agent.manager.DirectAgentAttache$Task.runInContext(DirectAgentAttache.java:315)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
Caused by: SR_BACKEND_FAILURE_47The SR is not available [opterr=The SR is not available [opterr=no such volume group: VG_XenStorage-fff6f0d5-aba6-601f-2272-566c63427a1a]]
    at com.xensource.xenapi.Types.checkResponse(Types.java:2021)
    at com.xensource.xenapi.Connection.dispatch(Connection.java:395)
    at com.cloud.hypervisor.xenserver.resource.XenServerConnectionPool$XenServerConnection.dispatch(XenServerConnectionPool.java:457)
    ... 24 more

syed commented:
Seems to be related to the resignature bit. Do you have the resignature pack installed on the XenServer @mike-tutkowski ? I will try to reproduce this locally today.

syed commented:
@mike-tutkowski I am able to reproduce this on my setup. I will debug this and let you know

syed commented:
@mike-tutkowski I've fixed the problem. I was not re-reading the resigned UUID and it was causing an error. I've fixed that and tested it with LVM based SR. It should work now.

syed commented:
I've added unit tests as well. Now If your CI gives a green, I will open this PR against the main cloudstack repo

mike-tutkowski commented:
Thanks, Syed!

I will continue to run my tests with your new code.

On Jun 29, 2016, at 9:46 AM, Syed Mushtaq Ahmed <notifications@github.com<mailto:notifications@github.com>> wrote:

I've added unit tests as well. Now If your CI gives a green, I will open this PR against the main cloudstack repo

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/syed/cloudstack/pull/4#issuecomment-229398849, or mute the threadhttps://github.com/notifications/unsubscribe/AC4SH9dyGAFuZ20WXVvu9_II2mqR0k3sks5qQpNngaJpZM4I2gkI.

mike-tutkowski commented:
Somehow this message of mine didn't get posted. I meant to post this before the stack trace I gave you to give you some context:

New info: I am unable to create a VM when using the resigning logic.

mike-tutkowski commented:
In any event, with your new code the resigning part works again.

Here are the results of TestVolumes.py:

test_00_check_template_cache (TestVolumes.TestVolumes) ... === TestName: test_00_check_template_cache | Status : SUCCESS ===
ok
Attach a volume to a stopped virtual machine, then start VM ... === TestName: test_01_attach_new_volume_to_stopped_VM | Status : SUCCESS ===
ok
Attach, detach, and attach volume to a running VM ... === TestName: test_02_attach_detach_attach_volume | Status : SUCCESS ===
ok
Attach volume to running VM, then reboot. ... === TestName: test_03_attached_volume_reboot_VM | Status : SUCCESS ===
ok
Detach volume from a running VM, then reboot. ... === TestName: test_04_detach_volume_reboot | Status : SUCCESS ===
ok
Detach volume from a stopped VM, then start. ... === TestName: test_05_detach_vol_stopped_VM_start | Status : SUCCESS ===
ok
Attach a volume to a stopped virtual machine, then start VM ... === TestName: test_06_attach_volume_to_stopped_VM | Status : SUCCESS ===
ok
Destroy and expunge VM with attached volume ... === TestName: test_07_destroy_expunge_VM_with_volume | Status : SUCCESS ===
ok
Delete volume that was attached to a VM and is detached now ... === TestName: test_08_delete_volume_was_attached | Status : SUCCESS ===
ok
Attach a data disk to a VM in one account and attach another data disk to a VM in another account ... === TestName: test_09_attach_volumes_multiple_accounts | Status : SUCCESS ===
ok
Attach more than one disk to a VM ... === TestName: test_10_attach_more_than_one_disk_to_VM | Status : SUCCESS ===
ok

---

Ran 11 tests in 2293.226s

OK

mike-tutkowski commented:
I am now re-running TestVolumes.py without using resigning. I can post those results once they're in.

syed commented:
This is good news @mike-tutkowski. Hopefully, there should be no problems without the resign code as it ran once without errors correct?

mike-tutkowski commented:
There shouldn't be any problems there, but I figured I'd re-run it to be on the safe side.

After TestVolumes.py is done, I can run TestSnapshots.py (this does volume snapshots in XenServer without resigning (copying), with resigning, then moving from an environment that didn't use resigning to one that does).

syed commented:
I see. The moving test should be interesting.

mike-tutkowski commented:
TestVolumes.py without resigning just completed. It was successful for all 11 tests.

test_00_check_template_cache (TestVolumes.TestVolumes) ... === TestName: test_00_check_template_cache | Status : SUCCESS ===
ok
Attach a volume to a stopped virtual machine, then start VM ... === TestName: test_01_attach_new_volume_to_stopped_VM | Status : SUCCESS ===
ok
Attach, detach, and attach volume to a running VM ... === TestName: test_02_attach_detach_attach_volume | Status : SUCCESS ===
ok
Attach volume to running VM, then reboot. ... === TestName: test_03_attached_volume_reboot_VM | Status : SUCCESS ===
ok
Detach volume from a running VM, then reboot. ... === TestName: test_04_detach_volume_reboot | Status : SUCCESS ===
ok
Detach volume from a stopped VM, then start. ... === TestName: test_05_detach_vol_stopped_VM_start | Status : SUCCESS ===
ok
Attach a volume to a stopped virtual machine, then start VM ... === TestName: test_06_attach_volume_to_stopped_VM | Status : SUCCESS ===
ok
Destroy and expunge VM with attached volume ... === TestName: test_07_destroy_expunge_VM_with_volume | Status : SUCCESS ===
ok
Delete volume that was attached to a VM and is detached now ... === TestName: test_08_delete_volume_was_attached | Status : SUCCESS ===
ok
Attach a data disk to a VM in one account and attach another data disk to a VM in another account ... === TestName: test_09_attach_volumes_multiple_accounts | Status : SUCCESS ===
ok
Attach more than one disk to a VM ... === TestName: test_10_attach_more_than_one_disk_to_VM | Status : SUCCESS ===
ok

---

Ran 11 tests in 2095.121s

OK

mike-tutkowski commented:
I've just kicked off TestSnapshots.py. This one takes a couple hours to run.

mike-tutkowski commented:
I have hit an issue where I cannot create a volume from a snapshot using the non-resigning logic.

Steps to reproduce:
1. Disable re-sign logic in DB.
2. Create a volume snapshot of a volume.
3. Try to create a volume from this volume snapshot.

ERROR [c.c.a.ApiAsyncJobDispatcher](API-Job-Executor-121:ctx-6c6cb467 job-899) (logid:1df58970) Unexpected exception while executing org.apache.cloudstack.api.command.admin.volume.CreateVolumeCmdByAdmin
com.cloud.utils.exception.CloudRuntimeException: Failed to create volume: 83
    at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:803)
    at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:175)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:106)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:161)
    at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:161)
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:91)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
    at com.sun.proxy.$Proxy214.createVolume(Unknown Source)
    at org.apache.cloudstack.api.command.admin.volume.CreateVolumeCmdByAdmin.execute(CreateVolumeCmdByAdmin.java:42)
    at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:150)
    at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:108)
    at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:554)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
    at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:502)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Failed to create volume from snapshot:Copy operation failed in 'StorageSystemDataMotionStrategy.handleCreateVolumeFromSnapshotBothOnStorageSystem'
    at org.apache.cloudstack.engine.orchestration.VolumeOrchestrator.createVolumeFromSnapshot(VolumeOrchestrator.java:406)
    at com.cloud.storage.VolumeApiServiceImpl.createVolumeFromSnapshot(VolumeApiServiceImpl.java:825)
    at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:775)
    ... 31 more

mike-tutkowski commented:
I only had a little time to perform debugging, but I can say that the problem is a NullPointerException is thrown from StorageSystemDataMotionStrategy.cacheSnapshotChain.

needCacheStorage returns true, but should return false in this case:

boolean needCacheStorage = needCacheStorage(snapshotInfo, volumeInfo);

performCopyOfVdi starts on line 1038:

```
private CopyCmdAnswer performCopyOfVdi(VolumeInfo volumeInfo, SnapshotInfo snapshotInfo, HostVO hostVO) {
    String value = _configDao.getValue(Config.PrimaryStorageDownloadWait.toString());
    int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));

    DataObject srcData = snapshotInfo;
    CopyCmdAnswer copyCmdAnswer = null;
    DataObject cacheData = null;
    boolean needCacheStorage = needCacheStorage(snapshotInfo, volumeInfo);

    if (needCacheStorage) {
        cacheData = cacheSnapshotChain(snapshotInfo, new ZoneScope(volumeInfo.getDataCenterId()));
        srcData = cacheData;

    }
```

mike-tutkowski commented:
Not sure if this is a good way to solve this problem, but I changed the code in needCacheStorage to the following and I was then able to create a volume from a volume snapshot in this scenario:

```
protected boolean needCacheStorage(DataObject srcData, DataObject destData) {
    DataTO srcTO = srcData.getTO();
    DataStoreTO srcStoreTO = srcTO.getDataStore();

    if (srcStoreTO instanceof PrimaryDataStoreTO || srcStoreTO instanceof NfsTO || srcStoreTO.getRole() == DataStoreRole.ImageCache) {
        return false;
    }

    DataTO destTO = destData.getTO();
    DataStoreTO destStoreTO = destTO.getDataStore();

    if (destStoreTO instanceof NfsTO || destStoreTO.getRole() == DataStoreRole.ImageCache) {
        return false;
    }

    if (LOGGER.isDebugEnabled()) {
        LOGGER.debug("needCacheStorage true, dest at " + destTO.getPath() + "; dest role " + destStoreTO.getRole().toString() + srcTO.getPath() + "; src role " +
            srcStoreTO.getRole().toString());
    }

    return true;
}
```

mike-tutkowski commented:
This is the main difference:

if (srcStoreTO instanceof PrimaryDataStoreTO

mike-tutkowski commented on engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:
```diff
@@ -280,7 +287,65 @@ private boolean usingBackendSnapshotFor(SnapshotInfo snapshotInfo) {
+    }
+
+    private Scope getZoneScope(Scope scope) {
+        ZoneScope zoneScope = null;
```
We don't need "= null" here.

mike-tutkowski commented on engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:
```diff
@@ -802,18 +1015,49 @@ private CopyCmdAnswer performResignature(DataObject dataObj, HostVO hostVO, bool
             _volumeService.grantAccess(volumeInfo, hostVO, volumeInfo.getDataStore());
 
-            Map<String, String> srcDetails = getSnapshotDetails(snapshotInfo);
+            Map<String, String> srcDetails = getSnapshotDetails((SnapshotInfo) snapshotInfo);
```
This cast is redundant.

mike-tutkowski commented:
With the changes to needCacheStorage that I put in, TestSnapshots.py came back successful:

test_01_create_volume_snapshot_using_sf_snapshot (TestSnapshots.TestSnapshots) ... === TestName: test_01_create_volume_snapshot_using_sf_snapshot | Status : SUCCESS ===
ok
test_02_create_volume_snapshot_using_sf_volume (TestSnapshots.TestSnapshots) ... === TestName: test_02_create_volume_snapshot_using_sf_volume | Status : SUCCESS ===
ok
test_03_create_volume_snapshot_using_sf_volume_and_sf_snapshot (TestSnapshots.TestSnapshots) ... === TestName: test_03_create_volume_snapshot_using_sf_volume_and_sf_snapshot | Status : SUCCESS ===
ok

---

Ran 3 tests in 7131.865s

OK

syed commented:
Nice! So I will fixup some of the stuff and open up the PR against the main Cloudstack repo. Thanks for the review and tests @mike-tutkowski  :)

syed commented on engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:
```diff
@@ -280,7 +288,72 @@ private boolean usingBackendSnapshotFor(SnapshotInfo snapshotInfo) {
+
+        // both snapshot and volume are on primary datastore. No need for a cache storage as
+        // hypervisor will copy directly
+        if (srcStoreTO instanceof PrimaryDataStoreTO && destStoreTO instanceof PrimaryDataStoreTO) {
```
@mike-tutkowski I've made this condition more explicit. Does this look good?

mike-tutkowski replied:
Will this cause a problem for creating a template from a volume snapshot that's on primary storage?

On Jun 30, 2016, at 11:36 AM, Syed Mushtaq Ahmed <notifications@github.com<mailto:notifications@github.com>> wrote:

In engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.javahttps://github.com/syed/cloudstack/pull/4#discussion_r69175635:

> @@ -280,7 +288,72 @@ private boolean usingBackendSnapshotFor(SnapshotInfo snapshotInfo) {
>          return Boolean.parseBoolean(property);
>      }
> -    private void handleCreateTemplateFromSnapshot(SnapshotInfo snapshotInfo, TemplateInfo templateInfo, AsyncCompletionCallback<CopyCommandResult> callback) {
> -    protected boolean needCacheStorage(DataObject srcData, DataObject destData) {
> -        DataTO srcTO = srcData.getTO();
> -        DataStoreTO srcStoreTO = srcTO.getDataStore();
> -        DataTO destTO = destData.getTO();
> -        DataStoreTO destStoreTO = destTO.getDataStore();
>   +
> -        // both snapshot and volume are on primary datastore. No need for a cache storage as
> -        // hypervisor will copy directly
> -        if (srcStoreTO instanceof PrimaryDataStoreTO && destStoreTO instanceof PrimaryDataStoreTO) {

@mike-tutkowskihttps://github.com/mike-tutkowski I've made this condition more explicit. Does this look good?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/syed/cloudstack/pull/4/files/0db8a41472f683ebd51c59d8ad4de1a19aeeb27d#r69175635, or mute the threadhttps://github.com/notifications/unsubscribe/AC4SH2ZZz9R5BTwYrpvV_EBHtkGf7Ssrks5qQ_3_gaJpZM4I2gkI.

syed replied:
I don't think so because in case of template, destStoreTO is not primarystorage
